File: examples/ios/Examples/MultipleUsers.m
//
//  MultipleUsers.m
//  RealmExamples
//
//  Created by Chris Bush on 2020-09-18.
//  Copyright © 2020 MongoDB, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "MyRealmApp.h"

// :code-block-start: link-identity-objc
@interface LinkIdentitiesExample : NSObject
@end

@implementation LinkIdentitiesExample {
    RLMApp *app;
    RLMUser *anonymousUser;
    // :hide-start:
    XCTestExpectation *expectation;
    // :hide-end:
}
// :hide-start:
- (instancetype)initWithExpectation:(XCTestExpectation *)expectation {
    self = [super init];
    if (self) {
        self->expectation = expectation;
    }
    return self;
}
// :hide-end:

// Entry-point for example.
- (void)runExample {
    app = [RLMApp appWithId:YOUR_REALM_APP_ID];
    [self logInAnonymously];
}

- (void)logInAnonymously {
    [app loginWithCredential:[RLMCredentials anonymousCredentials] completion:^(RLMUser *user, NSError *error) {
        if (error != nil) {
            NSLog(@"Failed to log in: %@", [error localizedDescription]);
            return;
        }

        // User uses app, then later registers an account
        [self registerNewAccountAsAnonymousUser: user];
    }];
}

- (void)registerNewAccountAsAnonymousUser:(RLMUser *)user {    
    NSString *email = @"link2@example.com";
    NSString *password = @"ganondorf";
    [[app emailPasswordAuth] registerUserWithEmail:email password:password completion:^(NSError *error) { 
            if (error != nil) {
                NSLog(@"Failed to register new account: %@", [error localizedDescription]);
                return;
            }

            // Successfully created account, now link it
            // with the existing anon user
            [self linkUser:self->anonymousUser withCredentials:[RLMCredentials credentialsWithEmail:email password:password]];     
    }];
}

- (void)linkUser:(RLMUser *)user withCredentials:(RLMCredentials *)credentials {
    [[app currentUser] linkUserWithCredentials:credentials completion:^(RLMUser *user, NSError *error) {
        if (error != nil) {
            NSLog(@"Failed to link user: %@", [error localizedDescription]);
            return;
        }

        NSLog(@"Successfully linked user: %@", user);
        
        // :hide-start:
        [self->expectation fulfill];
        // :hide-end:
    }];
}

@end 
// :code-block-end:

@interface MultipleUsers : XCTestCase

@end

@implementation MultipleUsers

- (void)tearDown {
    XCTestExpectation *expectation = [self expectationWithDescription:@"Remove anonymous user from device"];
    [[[MyRealmApp app] currentUser] removeWithCompletion:^(NSError *error) {
        [expectation fulfill];
    }];
    [self waitForExpectations:@[expectation] timeout:10.0];
}

- (void)testLinkIdentities {
    XCTestExpectation *expectation = [self expectationWithDescription:@"Link user completes"];
    [[[LinkIdentitiesExample alloc] initWithExpectation:expectation] runExample];
    [self waitForExpectations:@[expectation] timeout:10.0];
}

@end

File: examples/ios/Examples/MultipleUsers.swift
import XCTest
import RealmSwift

class MultipleUsers: XCTestCase {

    override func tearDown() {
        guard app.currentUser != nil else {
            return
        }
        let expectation = XCTestExpectation(description: "Remove anonymous user from device")
        app.currentUser!.remove { (error) in
            expectation.fulfill()
        }
        wait(for: [expectation], timeout: 3)
    }

    func testAddUser() {
        let joeExpectation = XCTestExpectation(description: "joe log in completes")
        let emmaExpectation = XCTestExpectation(description: "emma log in completes")
        // :code-block-start: add-user
        let app = App(id: YOUR_REALM_APP_ID)

        let joeCredentials = Credentials.emailPassword(email: "joe@example.com", password: "passw0rd")
        app.login(credentials: joeCredentials) { (result) in
            switch result {
            case .failure(let error):
                print("Login failed: \(error.localizedDescription)")
                // :hide-start:
                joeExpectation.fulfill()
                // :hide-end:
            case .success(let joe):
                // The active user is now Joe
                assert(joe == app.currentUser)
            }
        }

        let emmaCredentials = Credentials.emailPassword(email: "emma@example.com", password: "pa55word")
        app.login(credentials: emmaCredentials) { (result) in
            switch result {
            case .failure(let error):
                print("Login failed: \(error.localizedDescription)")
                // :hide-start:
                emmaExpectation.fulfill()
                // :hide-end:
            case .success(let emma):
                // The active user is now Joe
                assert(emma == app.currentUser)
            }
        }
        // :code-block-end:
        wait(for: [joeExpectation, emmaExpectation], timeout: 10)
    }

    func testListUsers() {
        // :code-block-start: list-users
        let app = App(id: YOUR_REALM_APP_ID)
        let users = app.allUsers
        users.forEach({ (key, user) in
            print("User: \(key) \(user)")
        })
        // :code-block-end:
    }

    func testSwitchUsers() {
        // :code-block-start: switch-user
        let app = App(id: YOUR_REALM_APP_ID)

        // :hide-start:
        XCTAssertNil(app.currentUser)
        // :hide-end:
        // ... log in ...
        // :hide-start:
        func getSomeOtherUser() -> User {
            var someOtherUser: User? = nil
            let expectation = XCTestExpectation(description: "it logs in")
            app.login(credentials: Credentials.anonymous) { (result) in
                switch result {
                case .failure(_):
                    XCTAssert(false)
                case .success(let user):
                    someOtherUser = user
                    user.logOut { (error) in
                        expectation.fulfill()
                    }

                }
            }
            wait(for: [expectation], timeout: 20)
            return someOtherUser!
        }
        // :hide-end:
        // Get another user on the device, for example with `app.allUsers`
        let secondUser: User = getSomeOtherUser()

        XCTAssertNotEqual(app.currentUser, secondUser)
        // assert(app.currentUser != secondUser)

        // Switch to another user
        //app.switch(to: secondUser)

        // The switch-to user becomes the app.currentUser
        //XCTAssertEqual(app.currentUser, secondUser)
        //assert(app.currentUser == secondUser)
        // :code-block-end:
    }

    func testLinkIdentities() {
        let expectation = XCTestExpectation(description: "Link user completes")

        // :code-block-start: link-identity
        let app = App(id: YOUR_REALM_APP_ID)

        func logInAnonymously() {
            app.login(credentials: Credentials.anonymous) { (result) in
                switch result {
                case .failure(let error):
                    print("Failed to log in: \(error.localizedDescription)")
                    // :hide-start:
                    XCTAssert(false, "Failed to log in: \(error.localizedDescription)")
                    // :hide-end:
                case .success(let user):
                    // User uses app, then later registers an account
                    registerNewAccount(anonymousUser: user)
                }
            }
        }

        func registerNewAccount(anonymousUser: User) {
            let email = "swift-link@example.com"
            let password = "ganondorf"
            app.emailPasswordAuth.registerUser(email: email, password: password) { (error) in
                guard error == nil else {
                    print("Failed to register new account: \(error!.localizedDescription)")
                    // :hide-start:
                    XCTAssert(false, "Failed to register new account: \(error!.localizedDescription)")
                    // :hide-end:
                    return
                }

                // Successfully created account, now link it
                // with the existing anon user
                link(user: anonymousUser, with: Credentials.emailPassword(email: email, password: password))
            }
        }

        func link(user: User, with credentials: Credentials) {
            user.linkUser(credentials: credentials) { (result) in
                switch result {
                case .failure(let error):
                    print("Failed to link user: \(error.localizedDescription)")
                    // :hide-start:
                    XCTAssert(false, "Failed to link user: \(error.localizedDescription)")
                    // :hide-end:
                case .success(let user):
                    print("Successfully linked user: \(user)")
                    // :hide-start:
                    expectation.fulfill()
                    // :hide-end:
                }
            }
        }

        logInAnonymously()
        // :code-block-end:
        wait(for: [expectation], timeout: 10)
    }
}
